### PR TITLE
Possibly fix race in StorageMemory.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix possible race in `StorageMemory`.
https://clickhouse-test-reports.s3.yandex.net/0/9cac8a7244063d2092ad25d45502611e18d3749c/stress_test_(thread)/stderr.log
Have no idea how to write a test.

### CODE REVIEW

File: src/Storages/StorageMemory.cpp [hunk 1]
```diff
@@ -31,7 +31,7 @@ class MemorySource : public SourceWithProgress
         : SourceWithProgress(metadata_snapshot->getSampleBlockForColumns(column_names_, storage.getVirtuals(), storage.getStorageID()))
         , column_names(std::move(column_names_))
         , begin(begin_)
-        , end(end_)
+        , end(end_) /// [begin, end]
```
Comment (alexey-milovidov) [inline src/Storages/StorageMemory.cpp:34]: If it's included, name them first, last.

Comment (alexey-milovidov) [reply src/Storages/StorageMemory.cpp:34]: Need a comment about the semantics of the `end` list iterator in presense of insertions.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
> Have no idea how to write a test.

Insert by single row in one thread.
Truncate in other thread periodically.
Select in third thread.

Should reproduce easily under TSan.

File: src/Storages/StorageMemory.cpp [hunk 2]
```diff
@@ -123,13 +128,18 @@ Pipes StorageMemory::read(
...
+        BlocksList::iterator first = data.begin();
+        BlocksList::iterator last = data.begin();
 
-        std::advance(begin, stream * size / num_streams);
-        std::advance(end, (stream + 1) * size / num_streams);
+        std::advance(first, stream * size / num_streams);
+        std::advance(last, (stream + 1) * size / num_streams);
 
-        pipes.emplace_back(std::make_shared<MemorySource>(column_names, begin, end, *this, metadata_snapshot));
+        if (first == last)
```
Comment (Akazz) [inline src/Storages/StorageMemory.cpp:137]: I do not think this case is possible at all:
 - loop body is only executed for `num_streams > 0`
 - Assume `R = size / num_streams`. `size >= num_streams` implies that `R >= 1`
 -  `stream + 1 > stream` implies `last > first` for any integral R >= 1

File: src/Storages/StorageMemory.cpp [hunk 3]
```diff
@@ -41,29 +43,32 @@ class MemorySource : public SourceWithProgress
...
     {
-        if (it == end)
+        if (is_finished)
         {
             return {};
         }
         else
         {
-            Block src = *it;
+            Block src = *first;
```
Comment (Akazz) [inline src/Storages/StorageMemory.cpp:52]: We could easily save a couple of smaller memory allocations by not copying Block (it has at least std::vector + std::unordered_map). Just dereference the iterator and memorize the rows_count ... I hope this makes sense

File: src/Storages/StorageMemory.cpp [hunk 4]
```diff
@@ -41,29 +43,32 @@ class MemorySource : public SourceWithProgress
...
+                ++first;
             return Chunk(std::move(columns), src.rows());
         }
     }
 private:
     Names column_names;
-    BlocksList::iterator begin;
-    BlocksList::iterator end;
-    BlocksList::iterator it;
+    BlocksList::iterator first;
```
Comment (Akazz) [inline src/Storages/StorageMemory.cpp:69]: current?

File: src/Storages/StorageMemory.cpp [hunk 5]
```diff
@@ -22,17 +22,19 @@ namespace ErrorCodes
 class MemorySource : public SourceWithProgress
 {
 public:
+    /// We use range [first, last] which includes right border.
+    /// It is needed because new elements may be added to list in other thread.
+    /// Appending of new element changes end() of std::list, which may cause data race otherwise.
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply src/Storages/StorageMemory.cpp:27]: It's not quite correct to say that the `end` is changing. It always points to the end of list.

File: src/Storages/StorageMemory.cpp (hunk 2 above)
Comment (KochetovNicolai) [reply src/Storages/StorageMemory.cpp:137]: I suppose you are right. But I wouldn't like calculating it reading this code.
This check costs nothing.

**Comment by KochetovNicolai** [comment]
> Should reproduce easily under TSan.

I cannot do it easily. Test I tried:
```bash
#!/usr/bin/env bash

CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
. "$CURDIR"/../shell_config.sh

set -e

$CLICKHOUSE_CLIENT -q "drop table if exists mem"
$CLICKHOUSE_CLIENT -q "create table mem (x UInt64) engine = Memory"

function f {
  for i in $(seq 1 100); do
    $CLICKHOUSE_CLIENT -q "select count() >= 0 from (select sleep(0.1) from mem settings max_block_size=1)" | grep -v '^1$' || true
  done
}

function g {
  for i in $(seq 1 1000); do
    $CLICKHOUSE_CLIENT -q "insert into mem values (1)"
  done
}

function h {
  for i in $(seq 1 1000); do
    $CLICKHOUSE_CLIENT -q "truncate table mem"
  done
}

f &
g &
h &
wait
```

No tsan crashes so far.

**Comment by Akazz** [comment]
I would suggest to try slightly different timings:
```
function f {
  for i in $(seq 1 100000); do
    $CLICKHOUSE_CLIENT -q "select count() >= 0 from (select * from mem settings max_block_size=1)" | grep -v '^1$' || true
  done
}

function g {
  for i in $(seq 1 10000); do
    $CLICKHOUSE_CLIENT -q "insert into mem values (1)"
    $CLICKHOUSE_CLIENT -q "insert into mem values (1)"
    $CLICKHOUSE_CLIENT -q "truncate table mem"
  done
}
```

**Pull request merged**